python/ray/data/_internal/execution/streaming_executor_state.py

```diff
@@ -0,0 +1,186 @@
+"""Contains classes that encapsulate streaming executor state.
+
+This is split out from streaming_executor.py to facilitate better unit testing.
+"""
+
+from typing import Dict, List, Optional
+
+import ray
+from ray.data._internal.execution.interfaces import (
+    RefBundle,
+    PhysicalOperator,
+)
+from ray.data._internal.execution.operators.input_data_buffer import InputDataBuffer
+from ray.data._internal.progress_bar import ProgressBar
+from ray.data._internal.stats import DatasetStats
+
+
+# Holds the full execution state of the streaming topology. It's a dict mapping each
+# operator to tracked streaming exec state.
+Topology = Dict[PhysicalOperator, "OpState"]
+
+
+class OpState:
+    """The execution state tracked for each PhysicalOperator."""
+
+    def __init__(self, op: PhysicalOperator):
+        # Each inqueue is connected to another operator's outqueue.
+        self.inqueues: List[List[RefBundle]] = [
+            [] for _ in range(len(op.input_dependencies))
+        ]
+        # The outqueue is connected to another operator's inqueue (they physically
+        # share the same Python list reference).
+        self.outqueue: List[RefBundle] = []
+        self.op = op
+        self.progress_bar = None
+        self.num_completed_tasks = 0
+        self.inputs_done_called = False
+
+    def initialize_progress_bar(self, index: int) -> None:
+        self.progress_bar = ProgressBar(
+            self.op.name, self.op.num_outputs_total() or 1, index
+        )
+
+    def num_queued(self) -> int:
+        return sum(len(q) for q in self.inqueues)
+
+    def num_active_tasks(self):
+        # TODO: optimize this?
+        return len(self.op.get_work_refs())
+
+    def add_output(self, ref: RefBundle) -> None:
+        self.outqueue.append(ref)
+        self.num_completed_tasks += 1
+        if self.progress_bar:
+            self.progress_bar.update(1)
+
+    def refresh_progress_bar(self) -> None:
+        if self.progress_bar:
+            queued = self.num_queued()
+            self.progress_bar.set_description(
+                f"{self.op.name}: {self.num_active_tasks()} active, {queued} queued"
+            )
+
+    def dispatch_next_task(self) -> None:
+        for i, inqueue in enumerate(self.inqueues):
+            if inqueue:
+                self.op.add_input(inqueue.pop(0), input_index=i)
+                return
+        assert False, "Nothing to dispatch"
+
+
+def build_streaming_topology(dag: PhysicalOperator) -> Topology:
+    """Build the streaming operator state topology for the given DAG.
+
+    This involves creating the operator state for each operator in the DAG,
+    registering it with this class, and wiring up the inqueues/outqueues of
+    dependent operator states.
+
+    Returns:
+        The topology dict holding the streaming execution state.
```

Comment:
We're relying on the insertion-order of the topology dictionary being a topological ordering of the DAG for progress bar initialization (and maybe operator selection for tie-breaks), right? If so, we should leave a comment somewhere indicating as much and a test covering it.